Fix warnings

sharwell opened the pull request:
- Fix improper/sub-optimal generics usage
- Call static method directly, not through instance
- Improved specification of compiler arguments in pom.xml

After this change, the project builds without warnings even with `-Xlint -Xlint:-serial`.

sharwell committed:
Use generics properly

sharwell committed:
Call static method directly

sharwell committed:
Updated compilerArgs; always build with -Xlint -Xlint:-serial

ndw commented:
Interesting. A useful refresher on some aspects of generics I've not used a lot. A couple of questions.
1. In places like `BaseURI.java`, what motivates the use of `<?>`. It doesn't seem to have any consequences elsewhere, though perhaps it avoids a compiler warning that I haven't noticed.
2. In Template.java, and a few other places, what motivated changing `Iterator<String> prefixes` to `Iterator<?> prefixes` and adding a `(String)` cast?

sharwell committed:
Fix additional build warnings due to generics usage

sharwell commented:
I rebased this change on commit 9976a1213374740bcb91e2d1a6e79b859506eba7, and added an additional commit to address warnings that were introduced by 2733096a0f1d3a599aa82cbc7d8c5a9b7f830f63 (a commit which was added after my original work on this branch).

sharwell commented:
To answer your questions:
1. `SequenceIterator` is a generic type. In these cases, we don't care what the generic type argument is so using `<?>` fixes the lint warnings without requiring changing a bunch of the code.
2. This is the most "interesting" change of the branch IMO, because it seems to be _removing_ generics. It turns out that `iteratePrefixes()` in the Saxon library returns `Iterator` (no generic argument specified), which can be implicitly assigned to `Iterator<String>`, but that introduces an `unchecked` warning. You can also assign it to `Iterator`, but that introduces a `rawtypes` warning. Assigning it to `Iterator<?>` is safe and does not result in any warnings, but like using a raw `Iterator` the `next()` method will require a cast to the target type.
